src/actions/actions.ts

```diff
@@ -67,4 +72,51 @@ export class Actions {
 
     return response;
   }
+
+  async deserialize({
+    payload,
+    sigHeader,
+    secret,
+    type,
+  }: {
+    payload: ActionPayload;
+    sigHeader: string;
+    secret: string;
+    type: 'authentication';
+  }): Promise<AuthenticationActionPayload>;
+
+  async deserialize({
+    payload,
+    sigHeader,
+    secret,
+    type,
+  }: {
+    payload: ActionPayload;
+    sigHeader: string;
+    secret: string;
+    type: 'user_registration';
+  }): Promise<UserRegistrationActionPayload>;
+
+  async deserialize({
+    payload,
+    sigHeader,
+    secret,
+    type,
+  }: {
+    payload: ActionPayload;
+    sigHeader: string;
+    secret: string;
+    type: 'authentication' | 'user_registration';
+  }) {
+    await this.verifyHeader({ payload, sigHeader, secret });
+
+    switch (type) {
+      case 'authentication':
+        return payload as AuthenticationActionPayload;
+      case 'user_registration':
+        return payload as UserRegistrationActionPayload;
+      default:
+        return unreachable(type);
+    }
+  }
```

Comment:
@faroceann just wondering why you took this approach with multiple method definitions?

Reply:
These overloads ensure the SDK consumer receives appropriate types based on the action type they use

Reply:
@faroceann Is there something here you think I should apply to #1179?